### PR TITLE
Add /sf dump command with blockstorage initially

### PR DESCRIPTION
**Pull request by Sefiraat**
## Description

Adds a new command /sf dump <blockstorage>. This command just collects the blocks held in BlockStorage into a single file denoting the item ID and the amount held in storage.

Recently many servers have had issues with BlockStorage and a simple fix (for some) has been to show them just how many of a certain block they have to allow them to triage and tackle the amount. This has been very positive for servers with ExoticGarden where they have had upwards of millions of several of plants which, when they purged them, released a not insignificant amount of RAM. 

Until now I have just offered this command as a separate JAR to server owners having issues and worked through the resolution with them but I think it makes sense to have the command here, regardless of the few use-cases.

## Proposed changes
Add a new command that dumps contents into a file.

image

image

image

image

## Checklist

- [x] I have fully tested the proposed changes and promise that they will not break everything into chaos.
- [x] I have also tested the proposed changes in combination with various popular addons and can confirm my changes do not break them.
- [ ] I have made sure that the proposed changes do not break compatibility across the supported Minecraft versions (1.14.* - 1.19.*).
- [x] I followed the existing code standards and didn't mess up the formatting.
- [ ] I did my best to add documentation to any public classes or methods I added.
- [ ] I have added `Nonnull` and `Nullable` annotations to my methods to indicate their behaviour for null values
- [ ] I added sufficient Unit Tests to cover my code.

### CODE REVIEW

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/BlockStorageCommand.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    protected String getDescription() {
+        return "commands.blockstorage.description";
+    }
+
+    @Override
+    public void onExecute(CommandSender sender, String[] args) {
+        if (sender.hasPermission("slimefun.command.blockstorage") || sender instanceof ConsoleCommandSender) {
+            final Map<String, Integer> map = new LinkedHashMap<>();
+            for (World world : Bukkit.getWorlds()) {
+                for (Config value : BlockStorage.getRawStorage(world).values()) {
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/BlockStorageCommand.java:42]: this is pretty expensive and we're also writing this on the main thread. Can we move this to another?

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/BlockStorageCommand.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+            try (BufferedWriter writer = new BufferedWriter(new FileWriter(path + "/BlockStorageSummary.txt"))) {
+                map.entrySet().stream().sorted(Map.Entry.comparingByValue()).forEach(
+                    entry -> {
+                        final String message = entry.getKey() + " -> " + entry.getValue();
+                        try {
+                            writer.write(message + "\n");
+                        } catch (IOException e) {
+                            e.printStackTrace();
+                        }
+                        // sender.sendMessage(message);
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/BlockStorageCommand.java:59]: remove?

**Review by WalshyDev — DISMISSED** [review]
I kinda feel like this should be under a `dump` command so we can add more to it...

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/BlockStorageCommand.java (hunk 2 above)
Comment (Sefiraat) [reply src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/BlockStorageCommand.java:59]: Nards

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/BlockStorageCommand.java (hunk 1 above)
Comment (Sefiraat) [reply src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/BlockStorageCommand.java:42]: Not sure if it helps, but I had this run on CYT at it's peak of blockstorage with like 30mil blocks and it still took ms to run, and it's one that's only due to run very situationally? Cant see why I couldn't run it async however

Comment (Sefiraat) [reply src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/BlockStorageCommand.java]: Both made async and moved into a dump subcommand. It means the completer only has one item for now, but it's at least future ready then.

**Comment by Sefiraat** [comment]
I'm not a fan of the tab completer, but I think that thing made need some work in a separate PR, imo.

**Review by J3fftw1 — CHANGES_REQUESTED** [review]
The file isnt stored in Slimefun right but in the folder Slimefun\
Also wouldnt a dump folder be even better?

**Comment by Sefiraat** [comment]
> The file isnt stored in Slimefun right but in the folder Slimefun\
> Also wouldnt a dump folder be even better?

Ah yeah it does make sense to have a dumps folder, and I think I’ll datetimestamp the file as well.

**Comment by Sefiraat** [comment]
YARGH

**Review by J3fftw1 — APPROVED** [review]
LGTM, even though me approving doesnt matter

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/SlimefunTabCompleter.java [hunk 3]
```diff
@@ -52,6 +52,12 @@ public List<String> onTabComplete(CommandSender sender, Command cmd, String labe
                 // Returning null will make it fallback to the default arguments (all online players)
                 return null;
             }
+        } else if (args[0].equalsIgnoreCase("dump")) {
+            // Only one option here for now, but preparing for further usages of /sf dump
+            List<String> suggestions = new LinkedList<>();
+            suggestions.add("blockstorage");
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/SlimefunTabCompleter.java:58]:
```suggestion
            List<String> suggestions = List.of("blockstorage");
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/DumpCommand.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+import java.util.function.Supplier;
+
+/**
+ * Provides commands to dump data for triaging
+ *
+ * @author Sefiraat
+ */
+class DumpCommand extends SubCommand {
+
+    private static final String PATH = Slimefun.instance().getDataFolder().getAbsolutePath() + "\\dumps\\";
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/DumpCommand.java:33]: This is a very Windows path
```suggestion
    private static final String PATH = Paths.get(Slimefun.instance().getDataFolder().getAbsolutePath(), "dumps");
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/DumpCommand.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+    @Override
+    protected String getDescription() {
+        return "commands.dump.description";
+    }
+
+    @Override
+    public void onExecute(CommandSender sender, String[] args) {
+        if (sender.hasPermission("slimefun.command.dump") || sender instanceof ConsoleCommandSender) {
+            String sub = args[1];
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/DumpCommand.java:48]: no args length check?

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/DumpCommand.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+                );
+            }
+        } else {
+            Slimefun.getLocalization().sendMessage(sender, "messages.no-permission", true);
+        }
+    }
+
+    @ParametersAreNonnullByDefault
+    private void createDump(CommandSender sender, String filePrefix, Supplier<List<String>> dump) {
+        final String dateTimeStamp = new SimpleDateFormat("'_'yyyyMMddHHmm'.txt'").format(new Date());
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/DumpCommand.java:62]: While I like to final all my vars it isn't standard in the SF codebase.

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/DumpCommand.java [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+        final String fullPath = PATH + filePrefix + dateTimeStamp;
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/DumpCommand.java:63]:
```suggestion
        final String fullPath = Paths.get(PATH, filePrefix + dateTimeStamp);
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/DumpCommand.java [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+            directory.mkdir();
+        }
+
+        try (BufferedWriter writer = new BufferedWriter(new FileWriter(fullPath))) {
+            writer.write("Dump created at: " + dateTimeStamp + "\n");
+            for (String string : dump.get()) {
+                writer.write(string + "\n");
+            }
+
+            Slimefun.getLocalization().sendMessage(
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/DumpCommand.java:76]: Technically this should be in a `finally`

**Review by WalshyDev — CHANGES_REQUESTED** [review]

**Comment by Sefiraat** [comment]
All done :)

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/DumpCommand.java [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+import java.util.function.Supplier;
+
+/**
+ * Provides commands to dump data for triaging
+ *
+ * @author Sefiraat
+ */
+class DumpCommand extends SubCommand {
+
+    private static final String PATH = Paths.get(Slimefun.instance().getDataFolder().getAbsolutePath(), "dumps").toString();
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/DumpCommand.java:35]: No reason for this to be static as this should be a singleton class.

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/DumpCommand.java [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+            Slimefun.getLocalization().sendMessage(
+                sender,
+                "commands.dump.complete",
+                true,
+                msg -> msg.replace("%path%", fullPath)
+            );
+        }
+    }
+
+    private List<String> dumpBlockStorage() {
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/DumpCommand.java:98]:
```suggestion
    private @Nonnull List<String> dumpBlockStorage() {
```

File: src/main/resources/languages/en/messages.yml [hunk 11]
```diff
@@ -39,6 +39,10 @@ commands:
     running: '&7Running debug mode with test: &6%test%'
     disabled: '&7Disabled debug mode.'
 
+  dump:
+    description: 'Contains commands to dump data to files'
```
Comment (TheBusyBiscuit) [inline src/main/resources/languages/en/messages.yml:43]: Might want to specify that it is a dev/debug tool
```suggestion
    description: 'Debugging tool to dump data to files'
```

**Review by TheBusyBiscuit — CHANGES_REQUESTED** [review]
Looks pretty good, I like it much better as a general `/sf dump` command that we can expand upon!

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/DumpCommand.java [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+        String dateTimeStamp = new SimpleDateFormat("'_'yyyyMMddHHmm'.txt'").format(new Date());
+        String fullPath = Paths.get(PATH, filePrefix + dateTimeStamp).toString();
+        File directory = new File(PATH);
+
+        if (!directory.exists()) {
+            directory.mkdir();
+        }
+
+        try (BufferedWriter writer = new BufferedWriter(new FileWriter(fullPath))) {
+            writer.write("Dump created at: " + dateTimeStamp + "\n");
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/DumpCommand.java:82]: Kinda weird to have `Dump created at: _20220101.txt`. This should probably be a more pretty date anyway

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/DumpCommand.java [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+import org.bukkit.World;
+import org.bukkit.command.CommandSender;
+import org.bukkit.command.ConsoleCommandSender;
+
+import javax.annotation.ParametersAreNonnullByDefault;
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.nio.file.Path;
```
Comment (Phoenix-Starlight) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/DumpCommand.java:18]:
```suggestion
```
Path doesn't seem to be needed.

**Review by Phoenix-Starlight — CHANGES_REQUESTED** [review]

**Comment by J3fftw1** [comment]
Messaged @Sefiraat privately if he still wants to work on it.
Closing it later if no response

**Pull request closed without merging**